x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -67,9 +79,48 @@
         Setting.Property.NodeScope
     );
 
-    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = Setting.byteSizeSetting(
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = new Setting<>(
         SHARED_CACHE_SETTINGS_PREFIX + "size",
-        ByteSizeValue.ZERO,
+        ByteSizeValue.ZERO.getStringRep(),
+        s -> ByteSizeValue.parseBytesSizeValue(s, SHARED_CACHE_SETTINGS_PREFIX + "size"),
+        new Setting.Validator<ByteSizeValue>() {
+
+            @Override
+            public void validate(final ByteSizeValue value) {
+
+            }
+
+            @Override
+            public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+                if (value.getBytes() > 0) {
+                    @SuppressWarnings("unchecked") final List<DiscoveryNodeRole> roles =
```

Comment:
I'd feel better if we had an assert that `roles` was not null so we fail fast, or if we want to be lenient (which we probably don't) do `settings.get(NodeRoleSettings.NODE_ROLES_SETTING, Collections.emptyList())`

Reply:
I don't think `roles` can be `null` here. We don't allow explicitly null-valued settings, since that represents unset, and therefore the default value.